File: 10026_적록색약.cpp
#include <iostream>
#include <queue>
using namespace std;
char box[101][101] = { 0 };
char cbox[101][101] = { 0 };
queue<int> xstr;
queue<int> ystr;

int i, j, N, x , y;
char mychar = 'Z';
int cal(char (*boxx)[101])
{
	int count = 0;
	for (i = 0; i < N; i++) {
		for (j = 0; j < N; j++) {
			x = i; y = j;
			if (boxx[i][j] != 'Z') {
				mychar = boxx[i][j];
				boxx[i][j] = 'Z';

				//Inqueue
				xstr.push(x);
				ystr.push(y);

				while (1) {
					bool nflag = false, eflag = false, sflag = false, wflag = false;
					bool key = false;

					x = xstr.front();
					y = ystr.front();
					//끝에 테두리에서 바깥으로 못나가게 하는 장치
					if (x == 0) {
						nflag = true;
					}
					if (x == N - 1) {
						sflag = true;
					}
					if (y == 0) {
						wflag = true;
					}
					if (y == N - 1) {
						eflag = true;
					}
					xstr.pop();
					ystr.pop();

					if (eflag == false && boxx[x][y + 1] != 'Z' && boxx[x][y + 1] == mychar) {
						boxx[x][y + 1] = 'Z';

						//Inqueue
						xstr.push(x);
						ystr.push(y + 1);
					}
					if (sflag == false && boxx[x + 1][y] != 'Z' && boxx[x + 1][y] == mychar) {
						boxx[x + 1][y] = 'Z';

						//Inqueue
						xstr.push(x + 1);
						ystr.push(y);
					}
					if (wflag == false && boxx[x][y - 1] != 'Z' && boxx[x][y - 1] == mychar) {
						boxx[x][y - 1] = 'Z';

						//Inqueue
						xstr.push(x);
						ystr.push(y - 1);
					}
					if (nflag == false && boxx[x - 1][y] != 'Z' && boxx[x - 1][y] == mychar) {
						boxx[x - 1][y] = 'Z';

						//Inqueue
						xstr.push(x - 1);
						ystr.push(y);
					}

					if (xstr.empty() == true && ystr.empty() == true)
					{
						count++;
						break;
					}
				}
			}
		}
	}
	return count;
}

int main(void){
	cin >> N;
	for (i = 0; i < N; i++){
		cin >> box[i];
	}	
	/*-----------------------*/
	//색맹전용칸 복사하기
	//적록색맹은 빨간색과 초록색을 구별 못하니까 초록색을 빨간색으로 퉁친다.
	for (i = 0; i < N; i++) {
		for (j = 0; j < N; j++) {
			if (box[i][j] == 'G') {
				cbox[i][j] = 'R';
			}
			else {
				cbox[i][j] = box[i][j];
			}
		}
	}
	int normal, abnormal;
	normal = cal(&box[0]);
	abnormal = cal(&cbox[0]);

	//출력
	printf("%d %d\n", normal, abnormal);
	return 0;
}

File: 10026_적록색약.py
#일단 값을 받아온다.
N = input()

#공간선언 및 빈 리스트 선언
Tbox = [[0] * 101 for i in range(101)]
box = [[0] * 101 for i in range(101)]
cbox = [[0] * 101 for i in range(101)]

# 100 X 100 배열 코드
TTT = [[0]*100]*100

xstr = []
ystr = []

def cal(boxx):
    count = 0

    for i in range(0,int(N),1):
        for j in range(0,int(N),1):
            x = i
            y = j
            if (boxx[i][j] != 'Z'):
                mychar = boxx[i][j]
                boxx[i][j] = "Z"

                xstr.append(x)
                ystr.append(y)
                
                while(True):
                    nflag,eflag,sflag,wflag = False
                    x = xstr[0]
                    y = ystr[0]
                    if(x==0):
                        nflag = True
                    if(x==int(N)-1):
                        sflag = True
                    if(y==0):
                        wflag = True
                    if(y==int(N)-1):
                        eflag = True
                    del xstr[0]
                    del ystr[0]

                    if(eflag==False and boxx[x][y+1]!='Z' and boxx[x][y+1] == mychar):
                        boxx[x][y+1] = 'Z'

                        xstr.append(x)
                        ystr.append(y+1)

                    if (sflag == False and boxx[x + 1][y] != 'Z' and boxx[x + 1][y] == mychar):
                        boxx[x + 1][y] = 'Z'

                        xstr.append(x + 1)
                        ystr.append(y)

                    if (wflag == False and boxx[x][y - 1] != 'Z' and boxx[x][y - 1] == mychar):
                        boxx[x][y - 1] = 'Z'

                        xstr.append(x)
                        ystr.append(y - 1)

                    if (nflag == False and boxx[x - 1][y] != 'Z' and boxx[x - 1][y] == mychar):
                        boxx[x - 1][y] = 'Z'

                        xstr.append(x - 1)
                        ystr.append(y)

                    if(not xstr and not ystr):
                        count = count + 1
                        break
    return count

#값 얻기
for i in range(0,int(N),1):
    Tbox[i] = input()

#input으로 값을 얻어오면 str형식으로 저장되어서 직접 하나하나 복사 때어온다.
for i in range(0,int(N),1):
    for j in range(0,int(N),1):
        box[i][j] = Tbox[i][j]

#색명전용 복사하기
for i in range(0,int(N),1):
    for j in range(0,int(N),1):
        if(box[i][j] == 'G'):
            cbox[i][j] = 'R'
        else:
            cbox[i][j] = box[i][j]

#값을 저장할 곳 및 출력
normal = cal(box)
abnormal = cal(cbox)
print("%d %d" %(normal,abnormal))